File: app/logging_module/logger.py
import sys
import json
import time
from typing import Any, Dict, Optional
from loguru import logger
from pythonjsonlogger import jsonlogger
from datetime import datetime


class StructuredLogger:
    _instance: Optional["StructuredLogger"] = None
    _initialized: bool = False

    def __new__(cls) -> "StructuredLogger":
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self, level: str = "INFO", format_type: str = "json"):
        if StructuredLogger._initialized:
            return
        
        self._configure_logger(level, format_type)
        StructuredLogger._initialized = True

    def _configure_logger(self, level: str, format_type: str):
        logger.remove()
        
        if format_type == "json":
            logger.add(
                sys.stdout,
                level=level,
                serialize=True,
                format=self._json_formatter,
                enqueue=True
            )
        else:
            logger.add(
                sys.stdout,
                level=level,
                format="<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | <level>{level: <8}</level> | <cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> - <level>{message}</level>",
                enqueue=True
            )

    def _json_formatter(self, record):
        log_record = {
            "timestamp": record["time"].strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
            "level": record["level"].name,
            "module": record["name"],
            "function": record["function"],
            "line": record["line"],
            "message": record["message"],
        }
        if record["extra"]:
            log_record["extra"] = record["extra"]
        if record["exception"]:
            log_record["exception"] = str(record["exception"])
        return json.dumps(log_record, ensure_ascii=False)

    def info(self, message: str, **kwargs: Any):
        logger.bind(**kwargs).info(message)

    def debug(self, message: str, **kwargs: Any):
        logger.bind(**kwargs).debug(message)

    def warning(self, message: str, **kwargs: Any):
        logger.bind(**kwargs).warning(message)

    def error(self, message: str, **kwargs: Any):
        logger.bind(**kwargs).error(message)

    def critical(self, message: str, **kwargs: Any):
        logger.bind(**kwargs).critical(message)

    def bind(self, **kwargs: Any) -> Any:
        return logger.bind(**kwargs)


def get_logger(name: Optional[str] = None) -> StructuredLogger:
    return StructuredLogger()

File: pkg/common/logger/logger.go
package logger

import (
	"log"
	"os"
)

type Logger struct {
	infoLogger  *log.Logger
	errorLogger *log.Logger
	debugLogger *log.Logger
	level       string
}

var defaultLogger *Logger

func init() {
	defaultLogger = NewLogger("info")
}

func NewLogger(level string) *Logger {
	return &Logger{
		infoLogger:  log.New(os.Stdout, "INFO: ", log.Ldate|log.Ltime|log.Lshortfile),
		errorLogger: log.New(os.Stderr, "ERROR: ", log.Ldate|log.Ltime|log.Lshortfile),
		debugLogger: log.New(os.Stdout, "DEBUG: ", log.Ldate|log.Ltime|log.Lshortfile),
		level:       level,
	}
}

func (l *Logger) Info(format string, v ...interface{}) {
	l.infoLogger.Printf(format, v...)
}

func (l *Logger) Error(format string, v ...interface{}) {
	l.errorLogger.Printf(format, v...)
}

func (l *Logger) Debug(format string, v ...interface{}) {
	if l.level == "debug" {
		l.debugLogger.Printf(format, v...)
	}
}

func (l *Logger) Warn(format string, v ...interface{}) {
	l.infoLogger.Printf("WARN: "+format, v...)
}

func Info(format string, v ...interface{}) {
	defaultLogger.Info(format, v...)
}

func Error(format string, v ...interface{}) {
	defaultLogger.Error(format, v...)
}

func Debug(format string, v ...interface{}) {
	defaultLogger.Debug(format, v...)
}

func Warn(format string, v ...interface{}) {
	defaultLogger.Warn(format, v...)
}
